### PR TITLE
feat(BentoServer): implement --reload option

### PR DESCRIPTION
**Pull request by parano**
## Checklist:

- [ ] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: bentoml/_internal/server/__init__.py [hunk 1]
```diff
@@ -71,20 +74,13 @@ def _start_ngrok_server() -> None:
 
 def serve_production(
     bento_path_or_tag: str,
-    working_dir: Optional[str] = None,
-    port: Optional[int] = None,
+    working_dir: t.Optional[str] = None,
+    app_workers: t.Optional[int] = None,
```
Comment (parano) [inline bentoml/_internal/server/__init__.py:78]: not related to this PR, I was thinking about how should we expose the `workers` option to end-user and thought it  would be helpful to name them a bit more specific, `api_worders` and `runner_workers`

File: bentoml/_internal/server/__init__.py [hunk 2]
```diff
@@ -71,20 +74,13 @@ def _start_ngrok_server() -> None:
...
+    runner_workers: t.Optional[int] = None,
+    port: t.Optional[int] = None,
 ) -> None:
-    if working_dir is not None:
-        working_dir = os.path.abspath(working_dir)
-
     svc = load(bento_path_or_tag, working_dir=working_dir)
 
-    import psutil
```
Comment (parano) [inline bentoml/_internal/server/__init__.py:82]: @bojiang we no longer need to limit `serve --production` to POSIX users right?  I think uvicorn have good support for windows

File: bentoml/_internal/server/__init__.py (hunk 1 above)
Comment (parano) [reply bentoml/_internal/server/__init__.py:78]: on a related note, `workers` should be limited to 1 for development mode

File: bentoml/_internal/server/__init__.py [hunk 3]
```diff
@@ -32,7 +35,7 @@ def serve_development(
     for _, runner in svc._runners.items():
         runner._setup()
 
-    if with_ngroxy:
```
Comment (parano) [inline bentoml/_internal/server/__init__.py:35]: also changed from `ngroxy` to `ngrok` cc @bojiang  I think the latter is the right name?

Comment (bojiang) [reply bentoml/_internal/server/__init__.py:35]: whops where is `ngroxy ` from lol

File: bentoml/_internal/server/__init__.py (hunk 2 above)
Comment (bojiang) [reply bentoml/_internal/server/__init__.py:82]: Theoretically, yes. I plan to reorganize windows serving when other parts' ready

File: bentoml/_internal/cli/bento_server.py [hunk 4]
```diff
@@ -35,19 +35,35 @@ def add_serve_command(cli) -> None:
...
+        type=click.BOOL,
+        is_flag=True,
+        help="Reload Service when code changes detected, this is only available in development mode.",
+        default=False,
+        show_default=True,
+    )
+    @click.option(
+        "--reload-delay",
+        type=click.FLOAT,
+        help="Delay between each check if the Service needs to be reloaded. Default is 0.25s.",
```
Comment (bojiang) [inline bentoml/_internal/cli/bento_server.py:49]:
```suggestion
        help="Delay in seconds between each check if the Service needs to be reloaded. Default is 0.25.",
```

File: bentoml/_internal/server/__init__.py (hunk 1 above)
Comment (bojiang) [reply bentoml/_internal/server/__init__.py:78]: IMO we should only expose this in the configuration file, allows users to config workers(or quota) for api server and each runner.

File: bentoml/_internal/service/service.py [hunk 5]
```diff
@@ -15,7 +15,7 @@
     from starlette.middleware import Middleware
     from starlette.types import ASGIApp
 
-WSGI_APP = t.NewType("WSGI_APP", object)
+WSGI_APP = t.Callable[[t.Callable, t.Mapping[str, t.Any]], t.Iterable[bytes]]
```
Comment (bojiang) [inline bentoml/_internal/service/service.py:18]: 👍 a typing_extension.Protocol is a little better here i guess.

**Pull request merged**